lnwallet/chancloser/rbf_coop_transitions.go

```diff
@@ -0,0 +1,1117 @@
+func (c *ClosingNegotiation) ProcessEvent(event ProtocolEvent, env *Environment,
+) (*CloseStateTransition, error) {
+
+	// There're two classes of events that can break us out of this state:
+	// we receive a confirmation event, or we receive a signal to restart
+	// the co-op close process.
+	switch msg := event.(type) {
+	// If we get a confirmation, then the spend request we issued when we
+	// were leaving the ChannelFlushing state has been confirmed.  We'll
+	// now transition to the StateFin state.
+	case *SpendEvent:
+		return &CloseStateTransition{
+			NextState: &CloseFin{
+				transitionEvent: msg,
+				ConfirmedTx:     msg.Tx,
+			},
+		}, nil
+
+	// Otherwise, if we receive a shutdown, or receive an event to send a
+	// shutdown, then we'll go back up to the ChannelActive state, and have
+	// it handle this event by emitting an internal event.
+	//
+	// TODO(roasbeef): both will have fee rate specified, so ok?
+	case *ShutdownReceived, *SendShutdown:
+		chancloserLog.Infof("ChannelPoint(%v): RBF case triggered, "+
+			"restarting negotiation", env.ChanPoint)
+
+		return &CloseStateTransition{
+			NextState: &ChannelActive{},
+			NewEvents: fn.Some(protofsm.EmittedEvent[ProtocolEvent]{
+				InternalEvent: fn.Some([]ProtocolEvent{event}),
+			}),
+		}, nil
+	}
+
+	// If we get to this point, then we have an event that'll drive forward
+	// the negotiation process.  Based on the event, we'll figure out which
+	// state we'll be modifying.
+	switch {
+	case c.PeerState.LocalState.ShouldRouteTo(event):
+		chancloserLog.Infof("ChannelPoint(%v): routing %T to local "+
+			"chan state", env.ChanPoint, event)
+
+		// Drive forward the local state based on the next event.
+		transition, err := c.PeerState.LocalState.ProcessEvent(
+			event, env,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		nextLocalState, ok := transition.NextState.(AsymmetricPeerState)
+		if !ok {
+			return nil, fmt.Errorf("expected %T to be "+
+				"AsymmetricPeerState", transition.NextState)
+		}
+
+		return &CloseStateTransition{
+			NextState: &ClosingNegotiation{
+				PeerState: DualPeerState{
+					LocalState:  nextLocalState,
+					RemoteState: c.PeerState.RemoteState,
+				},
+			},
+			NewEvents: transition.NewEvents,
+		}, nil
+
+	case c.PeerState.RemoteState.ShouldRouteTo(event):
+		chancloserLog.Infof("ChannelPoint(%v): routing %T to remote "+
+			"chan state", env.ChanPoint, event)
+
+		// Drive forward the remote state based on the next event.
+		transition, err := c.PeerState.RemoteState.ProcessEvent(
+			event, env,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		nextRemoteState, ok := transition.NextState.(AsymmetricPeerState)
+		if !ok {
+			return nil, fmt.Errorf("expected %T to be "+
+				"AsymmetricPeerState", transition.NextState)
+		}
+
+		return &CloseStateTransition{
+			NextState: &ClosingNegotiation{
+				PeerState: DualPeerState{
+					LocalState:  c.PeerState.LocalState,
+					RemoteState: nextRemoteState,
+				},
```

Comment:
Lotta boiler plate across these two. They almost look like they are the same block except for the state lense they operate on.

Reply:
Can consolidate.